clustering/train_clusterer.py

```diff
@@ -0,0 +1,162 @@
+import argparse
+import itertools
+
+import numpy as np
+
+import pickle
+from pathlib import Path
+
+import torch
+
+from tqdm.auto import tqdm
+
+from torch.utils.data import DataLoader, IterableDataset
+
+from kmeans_pytorch import KMeans as BalancedKMeans
+
+from transformers import pipeline
+
+from datasets import load_dataset
+
+from sklearn.manifold import TSNE
+from sklearn.decomposition import PCA
+
+import matplotlib.pyplot as plt
+
+from feature_extractor import FeatureExtractor
+from memmap_utils import np_memmap
+
+
+def load_model(path_to_model: Path):
+    with open(path_to_model, 'rb') as file:
+        output = pickle.load(file)
+
+        file.close()
+
+    return output
+
+
+def extract_features(corpus, feature_extractor, batch_size=32, max_chars=256):
+    corpus = [element[:max_chars] for element in corpus]
+    batches = np.array_split(corpus, len(corpus) // batch_size, axis=0)
+
+    features = []
+
+    for batch in tqdm(batches):
+        batch = list(batch) # batches is a list of numpy arrays
+
+        features_current = feature_extractor(batch)
+        features_current = np.max(features_current, axis=1)
+
+        features.append(features_current)
+
+    features = np.concatenate(features, axis=0)
+
+    return features
+
+
+def train_kmeans(features, n_clusters, path_to_kmeans, balanced=False, device='cpu'):
+    kmeans = BalancedKMeans(n_clusters=n_clusters, device=device, balanced=balanced)
```

Comment:
want to note that on my local, n_clusters is not a parameter, but num_clusters is -- does this work for you?

Reply:
My bad - I did some refactoring on Huu's codebase and changed its name, will update it ASAP

Reply:
additionally, since I'm not using KMeans, but kmeans -- cannot be instantiated without X, which needs to be a torch object or else it fails
if this worked for you, great -- will sync up with you to figure out why it didn't on my env
if not, then we'll need to figure out a way to use pytorch-kmeans in batches since instantiation isn't immediately possible -- open to suggestions here